internal/codespaces/grpc/client.go

```diff
@@ -0,0 +1,117 @@
+package grpc
+
+// gRPC client implementation to be able to connect to the gRPC server and perform the following operations:
+// - Start a remote JupyterLab server
+
+import (
+	"context"
+	"fmt"
+	"net"
+	"strconv"
+	"time"
+
+	"github.com/cli/cli/v2/internal/codespaces/grpc/jupyter"
+	"github.com/cli/cli/v2/pkg/liveshare"
+	"golang.org/x/crypto/ssh"
+	"google.golang.org/grpc"
+	"google.golang.org/grpc/credentials/insecure"
+	"google.golang.org/grpc/metadata"
+)
+
+const (
+	connectionTimeout = 5 * time.Second
+	requestTimeout    = 30 * time.Second
+)
+
+const (
+	codespacesInternalPort        = 16634
+	codespacesInternalSessionName = "CodespacesInternal"
+)
+
+type Client struct {
+	conn          *grpc.ClientConn
+	token         string
+	listener      net.Listener
+	jupyterClient jupyter.JupyterServerHostClient
+}
+
+type liveshareSession interface {
+	KeepAlive(string)
+	OpenStreamingChannel(context.Context, liveshare.ChannelID) (ssh.Channel, error)
+	StartSharing(context.Context, string, int) (liveshare.ChannelID, error)
+}
+
+// Connects to the gRPC server on the given port
+func Connect(ctx context.Context, session liveshareSession, token string) (*Client, error) {
+	listener, err := net.Listen("tcp", fmt.Sprintf("127.0.0.1:%d", 0))
+	if err != nil {
+		return nil, fmt.Errorf("failed to listen to local port over tcp: %w", err)
+	}
+
+	// Tunnel the remote gRPC server port to the local port
+	localGrpcServerPort := listener.Addr().(*net.TCPAddr).Port
+	internalTunnelClosed := make(chan error, 1)
+	go func() {
+		fwd := liveshare.NewPortForwarder(session, codespacesInternalSessionName, codespacesInternalPort, true)
+		internalTunnelClosed <- fwd.ForwardToListener(ctx, listener)
+	}()
+
+	// Attempt to connect to the given port
+	opts := []grpc.DialOption{
+		grpc.WithTransportCredentials(insecure.NewCredentials()),
+		grpc.WithBlock(),
+	}
+	ctx, _ = context.WithTimeout(ctx, connectionTimeout)
```

Comment:
```suggestion
	ctx, cancel = context.WithTimeout(ctx, connectionTimeout)
	defer cancel()
```

Reply:
Won't calling `defer cancel()` close the connection and cause calls to fail?

Reply:
It shouldn't, but this reminds me, should we write some tests?

Reply:
Since we're calling `grpc.Connect` and returning the client to then be used to call `StartJupyterServer` at a later time,  it seems like `defer cancel()` actually closes the connection before we can use it and we get the following error when we send the request: `failed to start JupyterLab server: failed to invoke JupyterLab RPC: rpc error: code = Unavailable desc = connection closed before server preface received`

And yes for tests! I've started working on them and was going to add them before merging the PR.